.travis.yml

```diff
@@ -127,82 +139,108 @@ linux_with_fuse: &linux_with_fuse
+        - docker build --rm -t travis_ci
+          --build-arg "TRAVIS_USER=$(id -un)"
+          --build-arg "TRAVIS_UID=$(id -u)"
+          --build-arg "TRAVIS_GROUP=$(id -gn)"
+          --build-arg "TRAVIS_GID=$(id -g)"
+          build-support/docker/travis_ci/
+        # Note that:
+        # * We mount ${HOME} to cache the ${HOME}/.cache/pants/rust-toolchain.
+        # * With no args ci.sh just bootstraps a pants.pex.
+        # * We also build fs_util, to take advantage of the rust code built during bootstrapping.
+        - docker run --rm -t
+          -v "${HOME}:/travis/home"
+          -v "${TRAVIS_BUILD_DIR}:/travis/workdir"
+          travis_ci:latest
+          sh -c "git clean -xfd && ./build-support/bin/ci.sh && ./build-support/bin/release.sh -f"
+        - aws --no-sign-request --region us-east-1 s3 cp ${TRAVIS_BUILD_DIR}/pants.pex ${BOOTSTRAPPED_PEX_URL_PREFIX}.linux
 
     # Build macOS engine
-    - name: "OSX Native Engine Binary Builder"
-      os: osx
+    - <<: *default_osx_config
+      <<: *native_engine_cache_config
+      name: "Build OSX native engine and pants.pex"
+      stage: *bootstrap
       # We request the oldest image we can (corresponding to OSX 10.11) for maximum compatibility.
       # We use 10.11 as a minimum to avoid https://github.com/rust-lang/regex/issues/489.
       # See: https://docs.travis-ci.com/user/reference/osx/#OS-X-Version
       osx_image: xcode8
+      env:
+        - CACHE_NAME=macospexbuild
+        - PREPARE_DEPLOY=1  # To deploy fs_util.
+      script:
+        # With no args ci.sh just bootstraps a pants.pex. We also build fs_util, to take advantage
+        # of the rust code built during bootstrapping.
+        - git clean -xfd && ./build-support/bin/ci.sh && ./build-support/bin/release.sh -f
+        - aws --no-sign-request --region us-east-1 s3 cp ${TRAVIS_BUILD_DIR}/pants.pex ${BOOTSTRAPPED_PEX_URL_PREFIX}.osx
+      after_failure:
+        - ./build-support/bin/ci-failure.sh
+
+    - <<: *default_linux_test_config
+      name: "Build Linux wheels"
       stage: *test
-      language: generic
       env:
+        - CACHE_NAME=linuxwheelsbuild
+        - RUN_PANTS_FROM_PEX=1
         - PREPARE_DEPLOY=1
       script:
-        - ./pants --version && ./build-support/bin/release.sh -n
+        - ./build-support/bin/release.sh -n
+
+    - <<: *default_osx_test_config
+      name: "Build OSX wheels"
+      osx_image: xcode8
+      env:
+        - CACHE_NAME=osxwheelsbuild
+        - RUN_PANTS_FROM_PEX=1
+        - PREPARE_DEPLOY=1
+      script:
+        - ./build-support/bin/release.sh -n
 
     # TODO: Update this to use 10.14 once it is available
-    - name: "OSX 10.12 Sanity Check"
-      <<: *osx_test_config
+    - <<: *default_osx_test_config
+      name: "OSX 10.12 sanity check"
       osx_image: xcode9.2
+      env:
+        - CACHE_NAME=macos10.12sanity
+      script:
+        - MODE=debug ./build-support/bin/travis-ci.sh -bm
```

Comment:
Follow up TODO: swap -b from "don't bootstrap" to "do bootstrap" so it's only present for the bootstrap shards (I'll happily do it, but just noting)

Reply:
If and when you do, note that the bootstrapping, and ONLY the bootstrapping, should run before we set `RUN_PANTS_FROM_PEX=1`. Any other logic must follow this line, or else that logic will likely end up bootstrapping again.